### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by starkmsu**

### CODE REVIEW

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs [hunk 1]
```diff
@@ -42,81 +60,199 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
...
+            var dates = new List<DateTime>();
+            for (DateTime start = from; start < to; start = start.AddHours(1))
             {
-                double tradeVolume =  await GetTradeVolumeAsync(
-                    start,
+                var tableName = GetTableName(baseAssetId, start);
+                if (existingTables.Contains(tableName))
+                    dates.Add(start);
+            }
+            await Task.WhenAll(dates.Select(d =>
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: Probably it is better to use batching here. If there are to many dates to AddTradeVolumeAsync may consume to many resources.

File: src/Lykke.Service.TradeVolumes/Controllers/TradeVolumesController.cs [hunk 2]
```diff
@@ -21,187 +21,197 @@ public TradeVolumesController(ITradeVolumesCalculator tradeVolumesCalculator)
...
-                _dateFormat,
-                CultureInfo.InvariantCulture,
-                DateTimeStyles.AssumeUniversal,
-                out DateTime date))
-                return StatusCode(
-                    (int)HttpStatusCode.BadRequest,
-                    ErrorResponse.Create($"DateStr parameter is mulformed - {_dateFormat} format is expected"));
-
-            double tradeVolume = await _tradeVolumesCalculator.GetPeriodAssetVolumeAsync(
+            return await GetPeriodClientAssetTradeVolume(
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes/Controllers/TradeVolumesController.cs]: You don't need here async/await. Just return GetPeriodClientAssetTradeVolume

File: src/Lykke.Service.TradeVolumes/Controllers/TradeVolumesController.cs [hunk 3]
```diff
@@ -21,187 +21,197 @@ public TradeVolumesController(ITradeVolumesCalculator tradeVolumesCalculator)
...
+        /// <param name="assetPairId">AssetPair Id</param>
+        /// <param name="fromDateStr">Start DateTime in yyyyMMddHH string format (Inclusive)</param>
+        /// <param name="toDateStr">Finish DateTime in yyyyMMddHH string format (Exclusive)</param>
+        [HttpGet("pair/{assetPairId}/all/{fromDateStr}/{toDateStr}")]
+        [SwaggerOperation("GetPeriodClientAssetPairTradeVolume")]
+        [ProducesResponseType(typeof(AssetPairTradeVolumeResponse), (int)HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> GetPeriodAssetPairTradeVolume(
+            string assetPairId,
+            string fromDateStr,
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes/Controllers/TradeVolumesController.cs]: Use DateTimeOffset or DateTime

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs [hunk 4]
```diff
@@ -42,81 +60,199 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
...
+                items = items.Where(i => i.PartitionKey != excludeClientId);
+            double result = items.Sum(i => i.BaseVolume.HasValue ? i.BaseVolume.Value : 0);
+            if (clientId == Constants.AllClients)
+            {
+                result /= 2;
+                var now = DateTime.UtcNow;
+                if (now.Subtract(date).TotalHours >= 1 || now.Hour != date.Hour)
+                {
+                    double quotingVolume = items.Sum(i => i.QuotingVolume.HasValue ? i.QuotingVolume.Value : 0);
+                    ThreadPool.QueueUserWorkItem(_ =>
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: Why not Task.Run()?

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs [hunk 5]
```diff
@@ -42,81 +60,199 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
...
+                token = response.ContinuationToken;
+                foreach (var table in response.Results)
+                {
+                    existingTables.Add(table.Name);
+                }
+            }
+            while (token != null);
+        }
+
+        private async void AddAllVolumeAsync(
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: An unhandled exception in this method can crash the application

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs [hunk 6]
```diff
@@ -42,81 +60,199 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
...
             return (baseTradeVolume, quotingTradeVolume);
         }
 
+        private async Task AddTradeVolumeAsync(
+            DateTime date,
+            string clientId,
+            string baseAssetId,
+            string quotingAssetId,
+            string excludeClientId,
+            List<double> tradeVolumes)
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: A ConcurrentBag class can be used here

**Review by MikhailGerasimov — CHANGES_REQUESTED** [review]

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs (hunk 6 above)
Comment (starkmsu) [reply src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: We don't need concurrent read operations here.

File: client/Lykke.Service.TradeVolumes.Client/Lykke.Service.TradeVolumes.Client.csproj [hunk 7]
```diff
@@ -15,6 +15,10 @@
     <PackageTags>Lykke</PackageTags>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
```
Comment (MikhailGerasimov) [inline client/Lykke.Service.TradeVolumes.Client/Lykke.Service.TradeVolumes.Client.csproj]: No configuration for Release

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs (hunk 6 above)
Comment (MikhailGerasimov) [reply src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: It also works with a concurrent write with less overhead than _lock.WaitAsync(). Plus the code will be clearer.

File: src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs [hunk 8]
```diff
@@ -42,81 +60,211 @@ public TradeVolumesRepository(IReloadingManager<string> connectionStringManager,
...
             {
                 var rowFilter = TableQuery.GenerateFilterCondition(_rowKey, QueryComparisons.Equal, quotingAssetId);
                 filter = TableQuery.CombineFilters(filter, TableOperators.And, rowFilter);
             }
             var query = new TableQuery<TradeVolumeEntity>().Where(filter);
             var items = await storage.WhereAsync(query);
-            return items.Sum(i => i.BaseVolume.HasValue ? i.BaseVolume.Value : 0);
+            if (clientId != excludeClientId && excludeClientId != null)
+                items = items.Where(i => i.PartitionKey != excludeClientId);
+            double result = items.Sum(i => i.BaseVolume.HasValue ? i.BaseVolume.Value : 0);
```
Comment (MikhailGerasimov) [inline src/Lykke.Service.TradeVolumes.AzureRepositories/TradeVolumesRepository.cs]: ?? is shorter

**Review by MikhailGerasimov — APPROVED** [review]

**Review by NovichikhinAlexey — APPROVED** [review]

**Pull request merged**